src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -260,6 +248,28 @@ private IISAwaitable DoFlushAsync()
 
         public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationToken = default(CancellationToken))
         {
+            // To avoid concurrent reading and writing, if we have a pending read,
+            // we must cancel it. TODO check if we need to lock here. I don't think we do because cancel_io will
+            // noop if there is no data to write, and OnAsyncCompletion should only be triggered once per read/write.
+            if (!_wasUpgraded)
+            {
+                lock (_readCancelSync)
+                {
+                    switch (_currentOperationType)
+                    {
+                        case CurrentOperationType.Read:
+                            NativeMethods.http_cancel_io(_pInProcessHandler);
+                            break;
+                        case CurrentOperationType.Write:
+                        case CurrentOperationType.Flush:
+                        case CurrentOperationType.None:
+                        default:
+                            _allowedToRead = false;
```

Comment:
Isn't it important that the data gets written to the PipeWriter before releasing the lock?

Because I think this is important, I also think the lock shouldn't be acquired until after InitializeResponseAwaited is called if it is called. If you do this, you probably won't need to reorder InitializeResponseAwaited and Output.WriteAsync below.

Reply:
A bit confused by this comment.

> Isn't it important that the data gets written to the PipeWriter before releasing the lock?

Are you suggesting I do something like:
```c#
public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationToken = default(CancellationToken))
{
    lock (_readCancelSync)
    {
        // do cancellations
        ....
        return Operation.WriteAsync();
    }
}

```

For the WriteAsyncAwaited case, do we do something like:

```c#
public async Task WriteAsyncAwaited(ArraySegment<byte> data, CancellationToken cancellationToken)
{
    // WriteAsyncAwaited is only called for the first write to the body.
    // Ensure headers are flushed if Write(Chunked)Async isn't called.
    await InitializeResponseAwaited(data.Count);
    if (!_wasUpgraded)
    {
        lock (_readCancelSync)
        {
            // Do cancellations
        }
    }
    await Output.WriteAsync(data, cancellationToken: cancellationToken);
}

```

Reply:
Close. In the second one I would still call WriteAsync inside the lock, but store the task and await it outside the lock.